src/merkle/mmr/partial.rs

```diff
@@ -188,6 +188,67 @@ impl PartialMmr {
     // STATE MUTATORS
     // --------------------------------------------------------------------------------------------
 
+    /// Adds a new peak and optionally track it.
+    ///
+    /// # Arguments
+    ///
+    /// When `track` is `true` the new leaf is tracked.
+    ///
+    /// # Returns
+    ///
+    /// A list of the new elements inserted to the [PartialMmr].
+    pub fn add(&mut self, leaf: RpoDigest, track: bool) -> Vec<(InOrderIndex, RpoDigest)> {
```

Comment:
I'm not sure the algorithm in this function covers all cases. Specifically, consider partial MMR with 7 leaves such that we track authentication path to leaf 5. If we add a new leaf using `PartialMmr::add()`, we won't insert new authentication nodes, and so, the data structure will become corrupted. Here is a test case that illustrates this:
```Rust
// build MMR with 7 leaves
let mut mmr = Mmr::default();
for el in (0..7).map(int_to_node) {
    mmr.add(el);
}

// derive a partila MMR from it which tracks authentication path to leaf 5
let mut partial_mmr = PartialMmr::from_peaks(mmr.peaks(mmr.forest()).unwrap());
let path_to_5 = mmr.open(5, mmr.forest()).unwrap().merkle_path;
let leaf_at_5 = mmr.get(5).unwrap();
partial_mmr.track(5, leaf_at_5, &path_to_5).unwrap();

// add a new leaf to both MMR and partial MMR
let leaf_at_7 = int_to_node(7);
mmr.add(leaf_at_7);
partial_mmr.add(leaf_at_7, false);

// the openings should be the same, but currently this fails
assert_eq!(mmr.open(5, mmr.forest()).unwrap(), partial_mmr.open(5).unwrap().unwrap());
```

I'm not sure how much more complicated the algorithm here needs to get to handle such cases, but as an alternative, we could just rely on the `apply()` method the way I did [here](https://github.com/0xPolygonMiden/miden-base/blob/main/objects/src/transaction/chain_mmr.rs#L98).

Reply:
Oh, nice caught! Thank you! I got the false impression things were working correctly because of the tests, but yeah, there was a massive mistake in the code! Fixing it.